exla/lib/exla/mlir/value.ex

```diff
@@ -1,320 +1,388 @@
+    result_types = typespecs_to_mlir_types([typespec])
+    op(func, "stablehlo.clamp", [min, operand, max], result_types) |> hd()
   end
 
   def select(
         %Value{function: func} = pred,
         %Value{function: func} = on_true,
-        %Value{function: func} = on_false
+        %Value{function: func} = on_false,
+        typespec
       ) do
-    ref =
-      EXLA.NIF.mlir_select(func.ref, pred.ref, on_true.ref, on_false.ref)
-      |> unwrap!()
+    result_types = typespecs_to_mlir_types([typespec])
+    op(func, "stablehlo.select", [pred, on_true, on_false], result_types) |> hd()
+  end
 
-    %Value{ref: ref, function: func}
+  def pad(
+        %Value{function: func} = operand,
+        %Value{function: func} = pad,
+        padding_config,
+        typespec
+      ) do
+    result_types = typespecs_to_mlir_types([typespec])
+
+    {padding_low, padding_high, padding_mid} = unzip_padding_config(padding_config)
+
+    attributes = [
+      edge_padding_low: attr_dense_i64_elements(padding_low),
+      edge_padding_high: attr_dense_i64_elements(padding_high),
+      interior_padding: attr_dense_i64_elements(padding_mid)
+    ]
+
+    op(func, "stablehlo.pad", [operand, pad], result_types, attributes: attributes) |> hd()
   end
 
-  def pad(%Value{function: func} = operand, %Value{function: func} = pad, padding_config) do
-    {padding_low, padding_high, padding_mid} =
-      Enum.reduce(padding_config, {[], [], []}, fn {low, high, mid},
-                                                   {low_acc, high_acc, mid_acc} ->
-        {[low | low_acc], [high | high_acc], [mid | mid_acc]}
-      end)
+  defp unzip_padding_config(padding_config),
+    do: unzip_padding_config(padding_config, {[], [], []})
 
-    ref =
-      EXLA.NIF.mlir_pad(
-        func.ref,
-        operand.ref,
-        pad.ref,
-        Enum.reverse(padding_low),
-        Enum.reverse(padding_high),
-        Enum.reverse(padding_mid)
-      )
-      |> unwrap!()
+  defp unzip_padding_config([], {low_acc, high_acc, mid_acc}) do
+    {Enum.reverse(low_acc), Enum.reverse(high_acc), Enum.reverse(mid_acc)}
+  end
 
-    %Value{ref: ref, function: func}
+  defp unzip_padding_config([{low, high, mid} | rest], {low_acc, high_acc, mid_acc}) do
+    unzip_padding_config(rest, {[low | low_acc], [high | high_acc], [mid | mid_acc]})
   end
 
-  def fft(%Value{function: func} = value, fft_kind, fft_length)
+  def fft(%Value{function: func} = value, fft_kind, fft_length, typespec)
       when fft_kind in [:fft, :ifft]
       when is_list(fft_length) or is_integer(fft_length) do
-    ref =
-      EXLA.NIF.mlir_fft(
-        func.ref,
-        value.ref,
-        if(fft_kind == :fft, do: 1, else: 0),
-        List.wrap(fft_length)
-      )
-      |> unwrap!()
+    result_types = typespecs_to_mlir_types([typespec])
 
-    %Value{value | ref: ref}
+    fft_type =
+      case fft_kind do
+        :fft -> attr_fft_type(:fft)
+        :ifft -> attr_fft_type(:ifft)
+      end
```

Comment:
```suggestion
    fft_type = attr_fft_type(fft_kind)
```

Reply:
Ah, I wanted to be explicit, but then added guards while refactoring, so that should be fine :)